### PR TITLE
FXML-1980: Add QuantLinear and DequantLinear ONNXToTosa

### PR DESCRIPTION
**Pull request by flemairen6**

### CODE REVIEW

File: src/Conversion/ONNXToTOSA/NN/DequantizeLinear.cpp [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+    // Quantization formula is (x - zero_point) * scale
+    // TODO : Axis attribute
+    TosaBuilder tosaBuilder(rewriter, op->getLoc());
+    OpAdaptor adaptor(operands, op->getAttrDictionary());
+    auto deqLinearOp = llvm::cast<ONNXDequantizeLinearOp>(op);
+    auto result = deqLinearOp.getResult().getType();
+
+    mlir::Value x = deqLinearOp.x();
```
Comment (maxbartel) [inline src/Conversion/ONNXToTOSA/NN/DequantizeLinear.cpp:46]:
```suggestion
    Value x = deqLinearOp.x();
```
and everywhere else

File: src/Conversion/ONNXToTOSA/NN/DequantizeLinear.cpp [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    auto x_scale = deqLinearOp.x_scale();
+    mlir::Value x_zero_point = deqLinearOp.x_zero_point();
+    auto loc = op->getLoc();
+
+    // Cast into the destination type first
+    auto castOp = tosa::CreateOpAndInfer<mlir::tosa::CastOp>(rewriter, loc, result, x).getResult();
```
Comment (maxbartel) [inline src/Conversion/ONNXToTOSA/NN/DequantizeLinear.cpp:52]:
```suggestion
    Value castOp = tosa::CreateOpAndInfer<mlir::tosa::CastOp>(rewriter, loc, result, x);
```
and for the others, too.
Looks a bit cleaner I think.

File: src/Conversion/ONNXToTOSA/NN/DequantizeLinear.cpp [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+//===------------- ONNXDequantizeLinearOp.cpp - ONNXDequantizeLinearOp --------------===//
+//
+// Copyright (c) 2022 Advanced Micro Devices, Inc.
+//
+// =============================================================================
+//
+// This file lowers ONNXDequantizeLinearOp operator to TOSA dialect.
+//
+//===----------------------------------------------------------------------===//
```
Comment (maxbartel) [inline src/Conversion/ONNXToTOSA/NN/DequantizeLinear.cpp:13]: NIT: Make those the same length (probably the shorter one)

File: src/Conversion/ONNXToTOSA/ONNXToTOSACommon.hpp [hunk 4]
```diff
@@ -149,7 +149,7 @@ namespace onnx_mlir {
 inline bool isTOSASignedInt(mlir::Type type) {
   mlir::IntegerType intType = type.dyn_cast<mlir::IntegerType>();
   std::set<unsigned> intWidth{8, 16, 32, 48, 64};
-  return intType && intType.isSignless() &&
+  return intType &&
```
Comment (maxbartel) [inline src/Conversion/ONNXToTOSA/ONNXToTOSACommon.hpp:152]: Why was this necessary? If the two new pattern have a special case, maybe we need to add a second type converter or check manually. I don't feel good just loosening the constraints here

File: src/Conversion/ONNXToTOSA/NN/QuantizeLinear.cpp
Comment (maxbartel) [inline src/Conversion/ONNXToTOSA/NN/QuantizeLinear.cpp:1]: The comments would be the same as in the other file. Maybe it makes sense to put them into a single file?

File: src/Conversion/ONNXToTOSA/NN/QuantizeLinear.cpp [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+#include "llvm/ADT/ArrayRef.h"
+#include <src/Dialect/Mlir/IndexExpr.hpp>
+
+using namespace mlir;
+
+namespace onnx_mlir {
+
+namespace {
+
+class ONNXQuantizeLinearOpLoweringToTOSA : public ConversionPattern {
```
Comment (maxbartel) [inline src/Conversion/ONNXToTOSA/NN/QuantizeLinear.cpp:32]: Why not use a `OpConversionPattern`? The adaptor would then be given as an argument.

File: src/Conversion/ONNXToTOSA/NN/QuantizeLinear.cpp [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    //    getTypeConverter()->convertType(qLinearOp.getResult().getType());
+
+    mlir::Value x = qLinearOp.x();
+    auto y_scale = qLinearOp.y_scale();
+    mlir::Value y_zero_point = qLinearOp.y_zero_point();
+    auto loc = op->getLoc();
+
+    // Replace the division by a reciprocal followed by a mul
+    auto recOp = tosa::CreateOpAndInfer<mlir::tosa::ReciprocalOp>(rewriter, loc, x.getType(), x).getResult();
+    auto mulOp = tosa::CreateOpAndInfer<mlir::tosa::MulOp>(rewriter, loc, x.getType(), recOp, y_scale, 0).getResult();
```
Comment (maxbartel) [inline src/Conversion/ONNXToTOSA/NN/QuantizeLinear.cpp:54]: An op to round is missing here. It is in the description in https://github.com/onnx/onnx/blob/main/docs/Operators.md#QuantizeLinear

File: test/mlir/conversion/onnx_to_tosa/NN/QuantizeLinear.mlir [hunk 7]
```diff
@@ -0,0 +1,17 @@
+// RUN: onnx-mlir-opt --shape-inference --convert-onnx-to-tosa -cse %s -split-input-file | FileCheck %s
+
+/// Test the default behavior of AveragePool with no padding
```
Comment (maxbartel) [inline test/mlir/conversion/onnx_to_tosa/NN/QuantizeLinear.mlir:3]:
```suggestion

```

File: test/mlir/conversion/onnx_to_tosa/NN/DequantizeLinear.mlir [hunk 8]
```diff
@@ -0,0 +1,16 @@
+// RUN: onnx-mlir-opt --shape-inference --convert-onnx-to-tosa -cse %s -split-input-file | FileCheck %s
+
+/// Test the default behavior of AveragePool with no padding
```
Comment (maxbartel) [inline test/mlir/conversion/onnx_to_tosa/NN/DequantizeLinear.mlir:3]:
```suggestion

```

File: src/Conversion/ONNXToTOSA/NN/QuantizeLinear.cpp
Comment (flemairen6) [reply src/Conversion/ONNXToTOSA/NN/QuantizeLinear.cpp:1]: I don't really have an opinion on whether they live in the same file or not. Do you know what is the common way of doing for upstream operator?

File: src/Conversion/ONNXToTOSA/NN/QuantizeLinear.cpp (hunk 6 above)
Comment (flemairen6) [reply src/Conversion/ONNXToTOSA/NN/QuantizeLinear.cpp:57]: Yes the documentation isn't great. I think I misunderstood the rounding for a clamping (which is done by the cast op)

File: test/mlir/conversion/onnx_to_tosa/NN/QuantizeLinear.mlir [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+  %1 = onnx.Constant dense<0> : tensor<i8>                                   
+  %2 = "onnx.QuantizeLinear"(%arg0, %0, %1) {axis = 1 : si64} : (tensor<32x3x224x224xf32>, tensor<f32>, tensor<i8>) -> tensor<32x3x224x224xi8>
+  "func.return"(%2) : (tensor<32x3x224x224xi8>) -> ()
+}
+// CHECK-LABEL:  @test_quantizeLinear(%arg0: tensor<32x3x224x224xf32>)
+// CHECK-DAG:    %[[SCALE:.*]] = "tosa.const"() {value = dense<3.125000e-02> : tensor<f32>}
+// CHECK-DAG:    %[[ZP:.*]] = "tosa.const"() {value = dense<0> : tensor<i8>} 
+// CHECK-DAG:    %[[REC:.*]] = "tosa.reciprocal"(%arg0) : (tensor<32x3x224x224xf32>)
+// CHECK-DAG:    %[[MUL:.*]] = "tosa.mul"(%[[REC]], %[[SCALE]]) {shift = 0 : i32} : (tensor<32x3x224x224xf32>, tensor<f32>)
+// CHECK-DAG:    %[[ADD:.*]] = "tosa.add"(%[[MUL]], %[[ZP]]) : (tensor<32x3x224x224xf32>, tensor<i8>)
```
[the remark this answers is not in the document]
Comment (flemairen6) [reply test/mlir/conversion/onnx_to_tosa/NN/QuantizeLinear.mlir]: I'm kinda surprised to see that, as there doesn't seem to be any assert or error triggering when lowering and executing that.
However this is an issue, should I create a constant tensor of the size of the input then? I was looking into it at first, but I've had trouble figuring out how to retrieve the value from the `zero_point` in onnx as it is an input and not an attribute.

File: src/Conversion/ONNXToTOSA/NN/QuantizeLinear.cpp (hunk 6 above)
Comment (flemairen6) [reply src/Conversion/ONNXToTOSA/NN/QuantizeLinear.cpp]: Now that I look at it a bit more, it seems that tosa only has `floor` and `ceil` as rounding operator, and nothing to do a more fine-grained rounding, which sounds wrong in all cases. Do you know if there have been discussions around this topic already?

Comment (flemairen6) [reply src/Conversion/ONNXToTOSA/NN/QuantizeLinear.cpp]: Looks like Matthias had the same questions https://discuss.mlplatform.org/t/tosa-op-with-round-half-to-even-lowering-pytorchs-quantize-per-tensor/165/13

Comment (maxbartel) [reply src/Conversion/ONNXToTOSA/NN/QuantizeLinear.cpp]: I read it in a different way (I am not 100% sure though), but let's discuss this tomorrow in the office

File: test/mlir/conversion/onnx_to_tosa/NN/QuantizeLinear.mlir (hunk 9 above)
[the remark this answers is not in the document]
Comment (maxbartel) [reply test/mlir/conversion/onnx_to_tosa/NN/QuantizeLinear.mlir]: You can either get the value from the `tosa.const` https://github.com/onnx/onnx-mlir/blob/375008b5b731bc47894bb4c5715951ef5e915570/src/Conversion/ONNXToTOSA/Math/ReduceMean.cpp#L60-L66 (because of the modules pass) or via the adaptor from the `onnx.const `(there is a helper function for this, too, I think). But I think you are right with creating a new const op

File: test/mlir/conversion/onnx_to_tosa/NN/DequantizeLinear.mlir [hunk 10]
```diff
@@ -0,0 +1,15 @@
+// RUN: onnx-mlir-opt --shape-inference --convert-onnx-to-tosa -cse %s -split-input-file | FileCheck %s
+
+func.func @test_dequantizeLinear(%arg0 : tensor<32x3x224x224xi8>) -> tensor<32x3x224x224xf32> {
+  %0 = onnx.Constant dense<3.125000e-02> : tensor<f32>                       
+  %1 = onnx.Constant dense<0> : tensor<i8>                                   
+  %2 = "onnx.DequantizeLinear"(%arg0, %0, %1) {axis = 1 : si64} : (tensor<32x3x224x224xi8>, tensor<f32>, tensor<i8>) -> tensor<32x3x224x224xf32>
```
[the remark this answers is not in the document]
Comment (flemairen6) [reply test/mlir/conversion/onnx_to_tosa/NN/DequantizeLinear.mlir:6]: As discussed offline, `ui8` is not supported by tosa/our constantOp lowering so it doesn't work.

File: src/Conversion/ONNXToTOSA/NN/QuantizeLinear.cpp (hunk 6 above)
Comment (flemairen6) [reply src/Conversion/ONNXToTOSA/NN/QuantizeLinear.cpp]: Discussed offline, cast seems to be doing the rounding properly

File: src/Conversion/ONNXToTOSA/NN/DequantizeLinear.cpp [hunk 11]
```diff
@@ -0,0 +1,84 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===------------- ONNXDequantizeLinearOp.cpp - ONNXDequantizeLinearOp --------------===//
+//
+// Copyright (c) 2022 Advanced Micro Devices, Inc.
```
Comment (maxbartel) [inline src/Conversion/ONNXToTOSA/NN/DequantizeLinear.cpp:7]: NIT: 2023

File: src/Conversion/ONNXToTOSA/NN/DequantizeLinear.cpp [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+    // constant, and create a constant of the same rank as the input out of it in order
+    // to have a correct sub.
+    mlir::ElementsAttr zeroPoint = tosa::getElementsAttrFromConst(x_zero_point);
+    auto zpValue = zeroPoint.getValues<int8_t>()[0];
+    llvm::SmallVector<int64_t, 4> tmpTensor;
+    for (uint i = 0; i < inputShape.size(); ++i) {
+      tmpTensor.emplace_back(1);
+    }
+    std::vector zpVec = std::vector<int8_t>{zpValue};
+    auto zpConst = tosaBuilder.getConst(zpVec, tmpTensor);
```
Comment (maxbartel) [inline src/Conversion/ONNXToTOSA/NN/DequantizeLinear.cpp:63]: NIT: You could also create a `tosaBuilder.getSplattedConst` that accepts int8 values. I am ok with this though.

File: src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.hpp [hunk 13]
```diff
@@ -48,6 +48,8 @@ T getValueFromTosaConst(mlir::Value &val) {
   return val.getDefiningOp<mlir::tosa::ConstOp>().getValue().cast<T>();
 }
 
+mlir::ElementsAttr getElementsAttrFromConst(mlir::Value &val);
```
Comment (maxbartel) [inline src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.hpp:51]: NIT: Maybe you can add a comment explaining that both onnx and tosa const ops are supported

File: test/mlir/conversion/onnx_to_tosa/NN/DequantizeLinear.mlir [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+  "func.return"(%2) : (tensor<32x3x224x224xf32>) -> ()
+}
+// CHECK-LABEL:  @test_dequantizeLinear(%arg0: tensor<32x3x224x224xi8>) -> tensor<32x3x224x224xf32>
+// CHECK-DAG:    %[[SCALE:.*]] = "tosa.const"() {value = dense<3.125000e-02> : tensor<f32>}
+// CHECK-DAG:    %[[ZP:.*]] = "tosa.const"() {value = dense<0> : tensor<1x1x1x1xi8>} : () -> tensor<1x1x1x1xi8>
+// CHECK-DAG:    %[[SUB:.*]] = "tosa.sub"(%arg0, %[[ZP]]) : (tensor<32x3x224x224xi8>, tensor<1x1x1x1xi8>) -> tensor<32x3x224x224xf32>
+// CHECK-DAG:    %[[CAST:.*]] = "tosa.cast"(%[[SUB]]) : (tensor<32x3x224x224xf32>) -> tensor<32x3x224x224xf32>
```
Comment (maxbartel) [inline test/mlir/conversion/onnx_to_tosa/NN/DequantizeLinear.mlir:13]: Why is the output of the sub a float? And the cast going from float to float?

File: test/mlir/conversion/onnx_to_tosa/NN/QuantizeLinear.mlir [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+  %0 = onnx.Constant dense<3.125000e-02> : tensor<f32>                       
+  %1 = onnx.Constant dense<0> : tensor<i8>                                   
+  %2 = "onnx.QuantizeLinear"(%arg0, %0, %1) {axis = 1 : si64} : (tensor<32x3x224x224xf32>, tensor<f32>, tensor<i8>) -> tensor<32x3x224x224xi8>
+  "func.return"(%2) : (tensor<32x3x224x224xi8>) -> ()
+}
+// CHECK-LABEL:  @test_quantizeLinear(%arg0: tensor<32x3x224x224xf32>) -> tensor<32x3x224x224xi8>
+// CHECK-DAG:    %[[SCALE:.*]] = "tosa.const"() {value = dense<3.125000e-02> : tensor<f32>}
+// CHECK-DAG:    %[[ZP:.*]] = "tosa.const"() {value = dense<0> : tensor<1x1x1x1xi8>} : () -> tensor<1x1x1x1xi8>
+// CHECK-DAG:    %[[REC:.*]] = "tosa.reciprocal"(%arg0) : (tensor<32x3x224x224xf32>) -> tensor<32x3x224x224xf32>
+// CHECK-DAG:    %[[MUL:.*]] = "tosa.mul"(%[[REC]], %[[SCALE]]) {shift = 0 : i32} : (tensor<32x3x224x224xf32>, tensor<f32>) -> tensor<32x3x224x224xf32>
```
Comment (maxbartel) [inline test/mlir/conversion/onnx_to_tosa/NN/QuantizeLinear.mlir:13]: This has the same broadcasting issue as the add. The ranks must match

File: test/mlir/conversion/onnx_to_tosa/NN/DequantizeLinear.mlir [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+// CHECK-DAG:    %[[MUL:.*]] = "tosa.mul"(%[[CAST]], %[[SCALE]]) {shift = 0 : i32} : (tensor<32x3x224x224xf32>, tensor<f32>) -> tensor<32x3x224x224xf32>
```
Comment (maxbartel) [inline test/mlir/conversion/onnx_to_tosa/NN/DequantizeLinear.mlir:14]: Same here

File: src/Conversion/ONNXToTOSA/NN/DequantizeLinear.cpp (hunk 12 above)
Comment (maxbartel) [reply src/Conversion/ONNXToTOSA/NN/DequantizeLinear.cpp:63]: Now that I see that `mul` (and basically all elementwise ops) also has this broadcasting issue, this makes even more sense

**Review by maxbartel — APPROVED** [review]
LGTM!

File: test/mlir/conversion/onnx_to_tosa/NN/QuantizeLinear.mlir [hunk 17]
```diff
@@ -0,0 +1,17 @@
+// RUN: onnx-mlir-opt --shape-inference --convert-onnx-to-tosa -cse %s -split-input-file | FileCheck %s
+
+func.func @test_quantizeLinear(%arg0 : tensor<32x3x224x224xf32>) -> tensor<32x3x224x224xi8> {
+  %0 = onnx.Constant dense<3.125000e-02> : tensor<f32>                       
+  %1 = onnx.Constant dense<0> : tensor<i8>                                   
+  %2 = "onnx.QuantizeLinear"(%arg0, %0, %1) {axis = 1 : si64} : (tensor<32x3x224x224xf32>, tensor<f32>, tensor<i8>) -> tensor<32x3x224x224xi8>
```
Comment (mgehre-amd) [inline test/mlir/conversion/onnx_to_tosa/NN/QuantizeLinear.mlir:6]: I'm wondering if this PR also supports quantization to unsigned integers? Maybe add a test case for that?

Related: TOSA has this odd thing where it uses signless integers to say "signed" (like `i8`) but unsigned integers to say "unsigned" (like `ui8`). Is this the same in onnx-mlir?

File: src/Conversion/ONNXToTOSA/NN/QuantizeLinear.cpp [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+    mlir::ElementsAttr zeroPoint = tosa::getElementsAttrFromConst(y_zero_point);
+    auto zpValue = zeroPoint.getValues<int8_t>()[0];
+    auto zpConst = tosaBuilder.getSplattedConst<int8_t>(zpValue, inputShape.size());
+    mlir::ElementsAttr scaleFactor = tosa::getElementsAttrFromConst(y_scale);
+    auto scaleFactorValue = scaleFactor.getValues<float>()[0];
+    auto scaleFactorConst = tosaBuilder.getSplattedConst<float>(scaleFactorValue, inputShape.size());
+    
+    // Quantization formula is ((x / y_scale) + y_zero_point)
+    // Replace the division by a reciprocal followed by a mul
+    Value recOp = tosa::CreateOpAndInfer<mlir::tosa::ReciprocalOp>(rewriter, loc, xType, x).getResult();
```
Comment (mgehre-amd) [inline src/Conversion/ONNXToTOSA/NN/QuantizeLinear.cpp:70]: Do you need `.getResult()` here and below? I think it works fine without those - then the calls return `mlir::tosa::ReciprocalOp` instead of `Value`, which CreateOpAndInfer also accepts as arguments.

**Review by mgehre-amd — APPROVED** [review]

File: test/mlir/conversion/onnx_to_tosa/NN/QuantizeLinear.mlir (hunk 17 above)
Comment (flemairen6) [reply test/mlir/conversion/onnx_to_tosa/NN/QuantizeLinear.mlir:6]: On paper it would probably work, but we don't have handling of the `ui8` type yet to my knowledge - I tried to add a test case according to @cmcgirr-amd comment and realised it was failing because the constants cannot be lowered

Comment (mgehre-amd) [reply test/mlir/conversion/onnx_to_tosa/NN/QuantizeLinear.mlir:6]: What exactly fails with ui8? Which pass does not handle ui8?

Would the ONNXQuantizeLinearOpLoweringToTOSA need to be changed to support is correctly? If yes, I suggest to add a check there and then not apply the transform.

Comment (cmcgirr-amd) [reply test/mlir/conversion/onnx_to_tosa/NN/QuantizeLinear.mlir:6]: Can you also add the case where it fails so we can track this. We will want this fixed in the future.

**Comment by maxbartel** [comment]
> > Hmm I thought that we could use the TOSA rescale op for some things, but it seems ONNX allows arbitrary scale values..
> 
> My understanding is that the TOSA rescale ops replaces a `quantize(dequantize(x))` pattern (where the scale factor of both ops are different - otherwise this is a noop) which appear between layers, e.g. `quantize - matmul - dequantize - quantize -matmul -dequantize` becomes `quantize - matmul -rescale - matmul -dequantize` Because `quantize(dequantize(x)` maps from int to int, the rescale only needs to support "fixed-point" multiply, right shift.
> 
> Is this what you were thinking about?

@mgehre-amd This is the best case scenario if everything is in integer space. But I also thought that we can convert a `QuantizeLinear` into something like `cast - rescale` and `DequantizeLinear` into `rescale - cast`. to get rid of all the arithmetic ops. This would have made matching also easier. Now that I read into both the TOSA and ONNX specification I don't think we can do it that way though

File: test/mlir/conversion/onnx_to_tosa/NN/QuantizeLinear.mlir (hunk 17 above)
Comment (maxbartel) [reply test/mlir/conversion/onnx_to_tosa/NN/QuantizeLinear.mlir:6]: The cast op in MLIR does in theory allow ui8 and ui16, but only if you look into the type definitions. The comment in https://github.com/llvm/llvm-project/blob/19732a3eaa978f2c65c233a97593b09e43a618cc/mlir/include/mlir/Dialect/Tosa/IR/TosaOps.td#L1743 does not mention it and the specification https://www.mlplatform.org/tosa/tosa_spec.html#_cast does not list it as supported.

**Comment by mgehre-amd** [comment]
> > > Hmm I thought that we could use the TOSA rescale op for some things, but it seems ONNX allows arbitrary scale values..
> > 
> > 
> > My understanding is that the TOSA rescale ops replaces a `quantize(dequantize(x))` pattern (where the scale factor of both ops are different - otherwise this is a noop) which appear between layers, e.g. `quantize - matmul - dequantize - quantize -matmul -dequantize` becomes `quantize - matmul -rescale - matmul -dequantize` Because `quantize(dequantize(x)` maps from int to int, the rescale only needs to support "fixed-point" multiply, right shift.
> > Is this what you were thinking about?
> 
> @mgehre-amd This is the best case scenario if everything is in integer space. But I also thought that we can convert a `QuantizeLinear` into something like `cast - rescale` and `DequantizeLinear` into `rescale - cast`. to get rid of all the arithmetic ops. This would have made matching also easier. Now that I read more into both the TOSA and ONNX specification I don't think we can do it that way though

Right, it's not that easy. You need to also turn the conv/matmul/... into integer. We do this in a pass called streamlining in FlexML-PL, and I will present our approach in todays meeting.

File: test/mlir/conversion/onnx_to_tosa/NN/QuantizeLinear.mlir (hunk 17 above)
Comment (cmcgirr-amd) [reply test/mlir/conversion/onnx_to_tosa/NN/QuantizeLinear.mlir:6]: Could we abuse the implementation of TOSA in MLIR for now to support unsigned? And push for the specification to change to support this? I would imagine this would also come with the push to support signed and unsigned integers in TOSA.

**Review by cmcgirr-amd — APPROVED** [review]
